### DISCUSSION TITLE
Leadfield Matrix from SimNIBS to be used for EEG calc

### DISCUSSION THREAD

**Discussion by michefabb** [simnibs/simnibs#408] [7.Creating head models] [answered]
Dear SimNIBS experts,

I am writing you to understand a point about the Leadfield Matrix created with SimNIBS and Brainstorm.
I ran charm on my subject using T1w and T2w (as well as the Freesurfer midthickness surface with 20484 vertices) and then ran `dwi2cond `using my subject DWI to create the conductivity tensors (as specified in [https://simnibs.github.io/simnibs/build/html/documentation/command_line/dwi2cond.html](https://github.com/simnibs/simnibs/discussions/url) ); then I ran the Leadfield Matrix calculation (using the "volume normalized" method) as shown in 
[https://simnibs.github.io/simnibs/build/html/tutorial/optimization.html#tutorial-leadfield](url)
[https://simnibs.github.io/simnibs/build/html/documentation/sim_struct/tdcsleadfield.html#tdcsleadfield-doc](url)

using the eeg cap Neuroelectrics (where I modified the Iz channel as "Electrode" and not "Fiducial" to have 64 channels) and using the options
```
tdcs_lf.anisotropy_type = 'vn'
tdcs_lf.tissues = [] 
```

Then with
```
from simnibs.eeg.forward import prepare_forward
forward=prepare_forward('/leadfield/sub-01_recon_leadfield_EEG10-10_Neuroelectrics_64ch.hdf5', True)
```
I got the Leadfield matrix for the forward EEG, however this is in the format 64 x 20484 x 3 for unconstrained dipole orientations  ... how can I transform to the format 64 x 20484 for orientations constrained perpendicularly with respect to the cortical surface as done in Brainstorm?

Appreciated!

Mic

**Comment by jduemose** [comment, marked as answer]
Hi, great that you figured it out yourself. I was about to suggest this exact function.

You need to project the matrix which is currently unit dipoles in [x,y,z] directions to the normal direction. You do this by taking the dot product with the surface normal. You can obtain surface normals for example like 

```
m=simnibs.mesh_tools.mesh_io.read_gifti_surface("/path/to/central/surface/gii")
normals = m.nodes_normals().value
```

**Reply by michefabb** [reply]
Thanks @jduemose ,

I did:
`from simnibs.mesh_tools import mesh_io
import numpy as np
central_surf_l=mesh_io.read_gifti_surface(tdcs_lf.subpath+'/surfaces/lh.central.gii')
central_surf_r=mesh_io.read_gifti_surface(tdcs_lf.subpath+'/surfaces/rh.central.gii')
central_surf_normals=np.concatenate((central_surf_l.nodes_normals().value,central_surf_r.nodes_normals().value))
constrained_forward=np.dot(forward['data'],central_surf_normals.T)`

however it goes out of memory! Did I something wrong?
ChatGTP suggests the following (which works in a moment):
`constrained_forward = np.einsum('ijk,jk->ij', forward['data'], central_surf_normals)`
but I am unsure if the above does what is intended...could you kindly help?

Thanks, appreciated,

Mic

**Reply by jduemose** [reply]
Yes, that is correct. I would probably just do `np.sum(forward["data"], central_surf_normals, axis=-1)` but it is the same thing.

You don't want the matrix multiplication functions (`dot`, `@`, ...) as they will dot *all* normals with all leadfields, hence your OOM error (i.e., you would get (64,20484,20484)). What you want is basically to dot lf[0,0] with normals[0], lf[0,1] with normals[1], lf[1,0] with normals[0], etc.

**Comment by michefabb** [comment]
Thanks @jduemose ! I tried with `constrained_forward = np.sum(forward["data"], central_surf_normals, axis=-1)` however this returns an error `TypeError: _sum_dispatcher() got multiple values for argument 'axis'`

Concerning the LF calculation, I ran it with the 'scalar' and the 'vn' method getting a LF Matrix very very similar, so I wished to doublecheck with you.

I ran the `dwi2cond` with also the Reverse Phase Encoding b0_u.nii.gz:
`dwi2cond --all --eddy --phasedir=y --readout=0.09027 sub-01_recon sub-01_dwi.nii.gz sub-01_dwi.bval sub-01_dwi.bvec b0_u.nii.gz`

this created in the folder `m2m_sub01-recon` a file `DTI_coregT1_tensor.nii.gz` and a subfolder `dMRI_prep`

then I did:

```
from simnibs import sim_struct, run_simnibs

tdcs_lf = sim_struct.TDCSLEADFIELD()


tdcs_lf.subpath = '/SimNIBS_ANISOTROPIC_dwi2cond/m2m_sub-01_recon'

tdcs_lf.pathfem = '/SimNIBS_ANISOTROPIC_dwi2cond/leadfield'

tdcs_lf.eeg_cap = '/SimNIBS_ANISOTROPIC_dwi2cond/m2m_sub-01_recon/eeg_positions/EEG10-10_Neuroelectrics_64ch.csv'

tdcs_lf.anisotropy_type = 'vn'

tdcs_lf.tissues = [] # no additional tissues (if not set, default [1006] is eyes which would create a LF Matrix with more than 20484 vertices)

run_simnibs(tdcs_lf)
```

Here https://simnibs.github.io/simnibs/build/html/documentation/sim_struct/tdcsleadfield.html#tdcsleadfield-doc [](url) it states that `tdcs_lf`

> Automatically finds the file d2c_subID/dti_results_T1space/DTI_conf_tensor.nii.gz based on fnamehead.

however this was not created (see above what was created by dwi2cond) and I am passing only `tdcs_lf.subpath` and not `tdcs_lf.fnamehead` . I thought maybe it is not reading the right file hence also the `'vn'` is not anisotropic (and hence gives a similar result to `'scalar'`); so I corrected adding the` fnamehead` and `fname_tensor`(could you clarify which is the correct file to be used, I assume `DTI_coregT1_tensor.nii.gz`):

```
tdcs_lf.fnamehead = tdcs_lf.subpath+'/sub-01_recon.msh'

tdcs_lf.fname_tensor = '/SimNIBS_ANISOTROPIC_dwi2cond/m2m_sub-01_recon/DTI_coregT1_tensor.nii.gz'
```

however again I got very very similar LF Matrices. Any other explanation?

A last question, do you know if it is possible to import the LF Matrix in Brainstorm? Or plot it with Gmsh?

Thanks, appreciated!

Mic

**Reply by jduemose** [reply]
Oh, my bad it should be `constrained_forward = np.sum(forward["data"] * central_surf_normals, axis=-1)` --- so *multiplicatication* not a comma!

As to BrainStorm, we don't provide functions to convert to their format (only fieldtrip and mne-python) as it is my impression that they have their own pipeline which basically calls simnibs. Perhaps if you do a little digging in their code you can find the functions they use for importing the results of the simulation into BS?

I don't have experience with anisotropy modeling, unfortunately. If you check the log of the simulation it should print something like `Using anisotropic volume normalized conductivities based on the file` when it reads that information. I will not expect big effects of anisotropy though. You could compare with results reported by others, e.g., Vorwerk 2014 "A guideline for head volume conductor modeling in EEG and MEG". You can calculate the RDM (defined in the article) for example between the two leadfields.

You can plot it with gmsh if you add it as "node data" to a `Msh` class (containing the source space). You could also export it to fieldtrip format. That will just give you some mat files which you can read with matlab and then plot it there (source space info and the forward solution is exported).

**Reply by michefabb** [reply]
Dear @[ ,

thanks for your usual quick and kind replies!

> Oh, my bad it should be constrained_forward = np.sum(forward["data"] * central_surf_normals, axis=-1) --- so multiplicatication not a comma!

It works, thanks!

> As to BrainStorm, we don't provide functions to convert to their format (only fieldtrip and mne-python) as it is my impression that they have their own pipeline which basically calls simnibs. Perhaps if you do a little digging in their code you can find the functions they use for importing the results of the simulation into BS?

I think BST imports the result of `charm` and simplifies to 5 tissues, but no way to import the results of `TDCSLeadfield` , I had to  recreate (see above) the unconstrained matrix and then recreate the matlab variable myself, then import.

> I don't have experience with anisotropy modeling, unfortunately. If you check the log of the simulation it should print something like Using anisotropic volume normalized conductivities based on the file when it reads that information. I will not expect big effects of anisotropy though. You could compare with results reported by others, e.g., Vorwerk 2014 "A guideline for head volume conductor modeling in EEG and MEG". You can calculate the RDM (defined in the article) for example between the two leadfields.

You are right, checking the log file, it was not reading the DTI, now it is. Here I am plotting the map of the constrained LF Matrix on the cortex after having imported in Brainstorm to compare the bottom result (FEM Isotropic with Duneuro) and the top result from SimNIBS TDCSLeadfield (using dwi2cond anisotropic 'vn' method).
![LF_SimNIBSAnisotropic_vs_DuneuroIso](https://github.com/user-attachments/assets/18c3ae4f-9ffc-4d0d-adaf-54dc3077e62a)
You can find more info at the post I created on BST forum [https://neuroimage.usc.edu/forums/t/creating-conductivity-with-simnibs-and-brainstorm/47423/11](https://neuroimage.usc.edu/forums/t/creating-conductivity-with-simnibs-and-brainstorm/47423/11).

I have a doubt I need to check with you: after the `prepare_forward` I checked the list of channels and it looks that the first channel changed from P2 (as it is in the EEG10-10_Neuroelectrics.csv file) to Cz and viceversa (channel 57 in the csv from Cz became P2)...I am wondering if there is some mismatch after calculating the average (maybe at [https://github.com/simnibs/simnibs/blob/master/simnibs/eeg/forward.py#L175](https://github.com/simnibs/simnibs/blob/master/simnibs/eeg/forward.py#L175)) or whatever...could I ask you to check and let me know please? I am also wondering the LF Matrix row 0 refers to which channel...I attach the list from the csv file and the prepare_forward result (ch_names).
[Ch_csv-prepare_forward.xlsx](https://github.com/user-attachments/files/16663298/Ch_csv-prepare_forward.xlsx)

I would also have another point where I would appreciate your opinion... I see it can be chosen wether to calculate the LF Matrix based on the electrical field or current density (pls correct if wrong):
if `field='E' ` the LF Matrix is in [V/(A*m)]  (actually it is V/m setting current to 1A, but for linearity this should be V/(A*m) )
if `field='J' ` the LF Matrix is in [1/(m^2)] 

Using `field='E' `  and multiplying the LF matrix by the matrix with the activity of the dipoles in [A*m] we get the [V] at the electrodes; however if I have the membrane potential in [V] of the vertices and wish to calculate the [V] at the electrode, this would be a much simpler "transfer function"...  any suggestion about how I could use the LF Matrix to calculate the V at electrodes?

Appreciated!

Mic

**Reply by jduemose** [reply]
I think you are okay with the electrodes. Apparently what happens during the simulation is that the reference electrode and the first electrode (index 0) are swapped meaning that the first electrode is always the reference in the leadfield file. I wasn't aware actually, however, the way I implemented the re-referencing should be agnostic to this as long as `ch_ref` matches `ch_names[0]`. Usually, the first channel is just used as reference for the simulations (hence `lf` will be `n_channels-1`) unless the EEG cap specifies a particular electrode as `ReferenceElectrode` in which case this should be used. The reference electrode is Cz in your case, hence Cz and P2 are swapped. So to recap: the order of the electrodes in the leadfield matrix should follow the order in `ch_names` returned by `prepare_forward`.

For calculating the EEG forward matrix using reciprocity as we do, you want to use `E` and not `J`. As to your last question: if you know V somewhere in the volume and want to calculate V elsewhere, I believe you would just use the known voltages as boundary conditions of your problem and solve (normally, what we do, is to iterate over all electrodes and set potential differences between each and the ref electrode - and then calculate E from there and so on). I am not sure what you want to use it for, however, if you look at the publications for the Wolters group for example, you will find that they talk a lot about 'source models'. In the St. venant approach you model a dipole as a set of nearby voltage differences which you obtain in such a way as to model your dipole optimally somehow. I imagine if you *knew* the potential distribution of each "dipole" you could use this to get a more accurate source model. I don't know if this answers your question.

Best,
Jesper

**Reply by michefabb** [reply]
Thanks @jduemose ,

very precious as always!
Concerning the last point: I would like to calculate the Leadfield Matrix where the rows are the channels and the columns the vertices of the midthickness cortex (exactly as it is done in output from `prepare_forward`), however the current LF Matrix must be multiplied by the value of the dipole [A*m] in each vertex to calculate the potential [V] at the electrode...I would like to multiply by the membrane potential in that vertex (this is what I get from my model) to calculate the potential at the electrode...any possibility to do it with SimNIBS (or to derive it from the current LF Matrix)?

Appreciated!

Mic

**Reply by jduemose** [reply]
I see. I guess you can use the same matrix if you are able to estimate the source strength and direction which corresponds to the potential (difference) you get from your model:
 
The primary source of the observed EEG signal (i.e., potentials far away from the brain sources) is believed to be post synaptic potentials in large groups of aligned neurons so if your observed potential difference can be said to represent such a cluster and you, for example, can assume two points between which this potential difference is observed then I suppose you could estimate the dipole (direction and moment) which has given rise to your observed field. Then you can multiply this with the LF you already have to estimate EEG potentials.

**Reply by michefabb** [reply]
Thanks @jduemose ! Very kind as always. If possible I would ask you also a last point.
Would you have any reference showing how to implement long term effects of TDCS/TMS in models?

Appreciated!

Mic

**Reply by michefabb** [reply]
Dear @jduemose and SimNIBS experts, any feedback about the last point?
Should I open a new post for this?

Appreciated!

Mic

**Reply by jduemose** [reply]
@michefabb I didn't even see it! But yes, I think that would be good. That way, it is easier for us to keep track of and assign to different members of the team :)

**Comment by xiongyy99255** [comment]


**Reply by michefabb** [reply]
Hi @xiongyy99255 , you can find above what I used, hope it is enough!
